### ISSUE TITLE
No serializable implementations of AWSCredentials

### ISSUE THREAD

**Issue by neil-rubens** [aws/aws-sdk-java#518]
There appears to be no serializable implementations of `AWSCredentials`; making it difficult to work with distributed framework e.g. Apache Spark.

**Comment by shorea** [comment]
Can you provide more details about your use case? Is this a blocker for using the SDK with Spark? How does Spark transport serializable objects like credentials securely within the cluster? Is there a way to get credentials in Spark via a more secure mechanism like instance profiles that don't require serializing the credentials object?

**Comment by neil-rubens** [comment]
Use Case: I am deploying Spark Cluster on EC2 (a popular and common option).  It is common to have multiple jobs running on a cluster (using different IAM profiles).

Since Apache Spark is distributed; all the object that it passes around need to be serializable.  One can just pass the authentication strings; but it would be a lot more appropriate to just pass the `AWSCredentials` object.

p.s. here is a brief discussion about serializable requirement of spark: https://forums.databricks.com/questions/369/how-do-i-handle-a-task-not-serializable-exception.html

**Comment by shorea** [comment]
Is it fair to say that you are passing request objects with contain the AWSCredentials?

**Comment by neil-rubens** [comment]
> Is it fair to say that you are passing request objects with contain the AWSCredentials?

Indeed I do.

p.s. spark machines are within the same VPC, so I assume it is secure.

**Comment by shorea** [comment]
Would it meet your needs to mark the credentials as transient and modify your spark task to restore the credentials from your local provider (probably InstanceProfileCredentialsProvider?)

**Comment by neil-rubens** [comment]
In my case this wouldn't quite work.  Here are brief elaborations as to why.

I have a Spark cluster that executes variety of jobs (w/ variety of credentials; which are currently passed in through a serializable class that I wrote).  Spark `slave` machines are added and destroyed on demand.  This setup works with serializable credentials.  p.s. there could be a potential security issues; if the credentials are passed over insecure channels

On the other hand; if I were to use `InstanceProfileCredentialsProvider` the keys would need to be propagated through all of the machines.  One of the challenges, is that if a new job (w/ new credentials) is submitted once the cluster is running; this would require propagating the credentials through all of the machines; similar procedure might need to be done whenever new machines are added to cluster.

**Comment by neil-rubens** [comment]
> Gotcha, so the master is the only instance with the Instance Profile attached and thus the only instance with access to credentials. Is that correct?

Yes; `master` is the only machine that has easy access to credentials.  Moreover; other credentials might need to be added to it at runtime.

> If we make the credentials transient though can you extract out the access key and secret key into your own serializable object and pass that around? If you are sure the serialized tasks are passed around securely that sounds like a viable option.

Actually; I would rather not have `transient` fields; I also do prefer for things to fail fast; so having transient fields might make it seem that things are working  even though it doesn't quite; making it particularly difficult to detect in distributed systems.  

Initially, I just made my own serializable credential class.  Later on; I ended up with a slightly more robust solution of storing credentials on S3; and retrieving them on demand (this does seem to allow decent security controls).  It is a somewhat hacky solution; probably it would be better if there is something more centralized. 

 Meanwhile, I do agree that making credentials serializable is a potential security issue; so it is probably beneficial not to have it; since am closing this issue.  Thank you for your help.
